src/queue.ts

```diff
@@ -0,0 +1,1052 @@
+                Type: "Task",
+                Resource: "arn:aws:states:::aws-sdk:sqs:sendMessageBatch",
+                Parameters: {
+                  QueueUrl: this.queueUrl,
+                  "Entries.$": heapVar,
+                },
+                Next: ASLGraph.DeferNext,
+              });
+            }
+          );
+        } else {
+          throw new SynthError(
+            ErrorCodes.Unexpected_Error,
+            `unknown DataType '${this.serializer.dataType}' is not supported by SQS by Step Functions`
+          );
+        }
+      },
+    });
+
+    this.receiveMessage = makeIntegration<
+      "AWS.SQS.ReceiveMessage",
+      (input: ReceiveMessageRequest) => Promise<ReceiveMessageResult<Message>>
+    >({
+      kind: "AWS.SQS.ReceiveMessage",
+      // asl: (call, context) => {},
+      native: {
+        bind: (func) => this.resource.grantConsumeMessages(func.resource),
+        preWarm: (context) => context.getOrInit(SQSClient),
+        call: async ([input], context) => {
+          const sqs: AWS.SQS = context.getOrInit(SQSClient);
+
+          const updatedQueueUrl =
+            lambdaQueueUrlRetriever?.(queueUrl) ?? queueUrl;
+
+          const response = await sqs
+            .receiveMessage({
+              ...input,
+              QueueUrl: updatedQueueUrl,
+            })
+            .promise();
+
+          return {
+            Messages: response.Messages?.map((message) => ({
+              ...message,
+              Message:
+                codec && message.Body
+                  ? codec.read(message.Body)
+                  : (message.Body as unknown as Message),
+            })),
+          };
+        },
+      },
+      asl: (call, context) => {
+        const queueUrl = this.queueUrl;
+        const serializer = this.serializer;
+        const input = call.args[0]?.expr;
+
+        this.resource.grantConsumeMessages(context.role);
+        const messagesHeap = context.newHeapVariable();
+
+        if (input === undefined) {
+          return context.evalContext(call, ({ addState }) => {
+            addState(receiveAndParse(undefined));
+            return {
+              jsonPath: messagesHeap,
+            };
+          });
+        } else {
+          return context.evalExprToJsonPathOrLiteral(
+            input,
+            (input, { addState }) => {
+              if (
+                ASLGraph.isJsonPath(input) ||
+                input.value === null ||
+                typeof input.value !== "object"
+              ) {
+                // Because of limitations in `Parameters` - need support for `Parameters.$`
+                throw new SynthError(
+                  ErrorCodes.Invalid_Input,
+                  `the 'input' to receiveMessage must be an object literal`
+                );
+              }
+              addState(receiveAndParse(input.value));
+              return {
+                jsonPath: messagesHeap,
+              };
```

Comment:
or you can put `output` on the SubState and return that.

```ts
return {
   stateState: "",
   output: { jsonPath: ... }
}
```

Reply:
(almost never use addState, but it does work how you are using it)

Reply:
I tried this and got a type error - not a valid SubState. Looks like our types don't properly capture the contract.

Reply:
I know it's missing `states` - that's not what I mean. If I set the return type to explicitly a `ASLGraph.SubState`, it doesn't allow the `output` property. That suggested to me that I can't use both at one time. Looks like that's not the case though.

Reply:
There is an `ASLGraph.OutputSubState`, which has the `output` property on it. `OutputSubState` and `OutputState` are `NodeResult`s. `SubState` and `State` are not. I wonder if it would be better to add an optional `state` property to `Output` instead of having an output on the `*State`.